sycl/include/CL/sycl/builtins.hpp

```diff
@@ -1543,17 +1543,7 @@ detail::enable_if_t<detail::is_genfloatf<T>::value, T> tan(T x) __NOEXC {
 } // __SYCL_INLINE_NAMESPACE(cl)
 
 #ifdef __SYCL_DEVICE_ONLY__
-#ifdef __GLIBC__
-extern "C" {
-extern SYCL_EXTERNAL void __assert_fail(const char *expr, const char *file,
-                                        unsigned int line, const char *func);
-}
-#elif defined(_WIN32)
-extern "C" {
-extern SYCL_EXTERNAL void _wassert(const wchar_t *wexpr, const wchar_t *wfile,
-                                   unsigned line);
-}
-#endif
+#include "devicelib/devicelib-assert.h"
```

Comment:
This approach has a limitation: since these wrappers are defined in the header file, it has to be #included in the translation unit that uses standard C and C++ functions. `CL/sycl.hpp` includes the necessary header, so most of the use-cases should work fine. 

If `CL/sycl.hpp` is not included and standard functions are used in device code (e.g. in functions defined as SYCL_EXTERNAL), user have to manually add `#include` directive or `-include` compiler option.

Reply:
Can we put the necessary parts to the compiler headers?
Something similar to https://github.com/intel/llvm/blob/sycl/clang/lib/Headers/__clang_cuda_libdevice_declares.h?

Reply:
We can, but why do we need to do this?

Reply:
To lift the limitation you described in this comment.

Reply:
The benefit of moving this to "compiler headers" is that the wrappers will be automatically engaged, when a standard header (e.g. assert.h) is included.

So the new assert.h would provide the implementation for __assert_fail, and then \#include_next the real assert.h.  We would need a new sub-directory in the compiler headers directory, e.g. device_wrappers, where we will put the new header files.

This will be transparent for programming models that do not require sycl.hpp, but also rely on the devicelib implementation.

At the same time, I am concerned about this change.  Let's say we do the same for FP64 math.  All wrappers will declare some functions with 'double' arguments, and the functions will remain in the final image at O0.  As I understand, we want to make FP64 usage conditional.  How do we do that with the header files implementation approach?

Reply:
1. I'm not sure if clang will emit functions not called from the device code.
2. AFAIK, translator do not translate "dead" declarations.

This should help with this problem, if I understand it correctly.

Reply:
Oh, I did not notice that they were 'inline'.  This should work then.

Reply:
And now I wonder if this would conflict with the non-inline declaration of __assert_fail in glibc's assert.h, which would matter, if we move to the compiler headers scheme (and use include_next).

Reply:
@vzakhari, so your suggestion is instead of having `devicelib-assert.h` and explicit `#include`, provide the wrapper in `assert.h` and `#include_next` the real `assert.h`. Is this accurate?

This does solve the problem I mentioned above. We have to make sure that the wrapper `assert.h` is first in include search path (not sure how, we probably need the driver support here).

Reply:
Yes, you got it right.  Yes, the driver support is needed to place -I\<clang\>/\<headers\>/devicelib_wrappers berore anything else.

Reply:
> This approach has a limitation: since these wrappers are defined in the header file, it has to be #included in the translation unit that uses standard C and C++ functions. `CL/sycl.hpp` includes the necessary header, so most of the use-cases should work fine.
> 
> If `CL/sycl.hpp` is not included and standard functions are used in device code (e.g. in functions defined as SYCL_EXTERNAL), user have to manually add `#include` directive or `-include` compiler option.

Maybe we can tell users to include <CL/sycl.hpp> if they want to use std:: in device code. This requirement can be added to document.

Reply:
> The benefit of moving this to "compiler headers" is that the wrappers will be automatically engaged, when a standard header (e.g. assert.h) is included.
> 
> So the new assert.h would provide the implementation for __assert_fail, and then #include_next the real assert.h. We would need a new sub-directory in the compiler headers directory, e.g. device_wrappers, where we will put the new header files.
> 
> This will be transparent for programming models that do not require sycl.hpp, but also rely on the devicelib implementation.
> 
> At the same time, I am concerned about this change. Let's say we do the same for FP64 math. All wrappers will declare some functions with 'double' arguments, and the functions will remain in the final image at O0. As I understand, we want to make FP64 usage conditional. How do we do that with the header files implementation approach?

Hi, I tried to enable cmath and cmath-fp64 in header file and found fp64 function will exist in final spirv file even though I am only running some float math tests. If we include both fp and fp64 headers, it may lead to problems on platform without fp64 support, just as Slava mentioned.

Reply:
> Oh, I did not notice that they were 'inline'. This should work then.

Oh, I didn't notice this and it should work with 'inline', I will verify.

Reply:
It may be OK to use "inline" as long as these files only provide some simple wrappers. This may become a compilation time issue, if they grow into something more. So I think we should not abandon the precompiled implementation right away. Does it make sense?

Reply:
Hi, Slava.
I am also considering this, I tried inline and it turned out to work for cmath functions as those functions are very simple wrapper. But I doubt whether it can work on some complicated implementation and we have a case currently, we need to enable some MSVC internal math functions in device library for supporting std::complex:
https://github.com/otcshare/llvm/pull/997
And some of those MSVC internal functions are much more complicated. I will verify this later.

Reply:
Ge, aren't those complex implementations supposed to reside in the SPV part of the library and the wrappers should be just calling __devicelib_* implementations in the SPV libs? I thought this was the design from the beginning..

Reply:
Hi, Slava.
libstdc++, libc++ and MSVC complex implementation are different:
libstdc++: based on c99 complex API, so we need to provide all c99 complex math functions in libdevice lib

libc++: all implementation are included in header file but depends on libm function, so we need to support cmath functions in devicelib to enable.

MSVC: all implementation are included in header file but depends on libm function and some MSVC internal math functions.

For functions included in SPV, we expect backend may provide some optimized version to override and we don't think backend would like to override those MSVC internal functions as they are only used by Windows.

Reply:
Thank you for the explanation, Ge! I will need to comprehend it tomorrow )

Reply:
> I tried inline and it turned out to work for cmath functions as those functions are very simple wrapper. But I doubt whether it can work on some complicated implementation and we have a case currently

AFAIR, `inline` does not force inlining, and a linked device program will have a single definition of an inline function. So this should be identical to having this function in an object file.

Reply:
> > I tried inline and it turned out to work for cmath functions as those functions are very simple wrapper. But I doubt whether it can work on some complicated implementation and we have a case currently
> 
> AFAIR, `inline` does not force inlining, and a linked device program will have a single definition of an inline function. So this should be identical to having this function in an object file.

Hi, Andrew.
I think it doesn't matter whether compiler decide to inline the functions or not. We split the device libraries into float and double to support devices without fp64 support. If we move all the implementation into header files, users will include both float and double math functions in their source code even they don't use double in device code and this may introduce unnecessary "double" code in their spirv files which will lead to failures in devices without fp64 support.
If we added "inline" to decorate the functions, I find the functions not invoked in device code will not exist in user's spirv file which seems to be able to solve the problem. But as we know, compiler may deny the "inline" if the functions are too complicated, I am not sure whether adding "inline" decoration can solve the problems  under such circumstances, I will try it later.
Thank you very much.

Reply:
>  If we move all the implementation into header files, users will include both float and double math functions in their source code even they don't use double in device code and this may introduce unnecessary "double" code in their spirv files which will lead to failures in devices without fp64 support.
> If we added "inline" to decorate the functions, I find the functions not invoked in device code will not exist in user's spirv file which seems to be able to solve the problem.

So basically we have the following:
```
// used in a kernel
inline float cosf(float f) {
 ... 
}

// not used
inline double cos(double d) {
  ...
}

void kernel() {
  cosf(...);
}
```
From my understanding, if a function is not called from a kernel, it is not emitted to device LLVM IR (regardless of whether it is inline or not).